stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -382,19 +380,32 @@ for Tri in (:UpperTriangular, :LowerTriangular)
     # 2 args
     for (fun, f) in zip((:*, :rmul!, :rdiv!, :/), (:identity, :identity, :inv, :inv))
         @eval $fun(A::$Tri, D::Diagonal) = $Tri($fun(A.data, D))
-        @eval $fun(A::$UTri, D::Diagonal) = $Tri(_set_diag!($fun(A.data, D), D.diag, $f))
+        @eval $fun(A::$UTri, D::Diagonal) = $Tri(_setdiag!($fun(A.data, D), $f, D.diag))
     end
     for (fun, f) in zip((:*, :lmul!, :ldiv!, :\), (:identity, :identity, :inv, :inv))
         @eval $fun(D::Diagonal, A::$Tri) = $Tri($fun(D, A.data))
-        @eval $fun(D::Diagonal, A::$UTri) = $Tri(_set_diag!($fun(D, A.data), D.diag, $f))
+        @eval $fun(D::Diagonal, A::$UTri) = $Tri(_setdiag!($fun(D, A.data), $f, D.diag))
     end
-    # 3 args
+    # 3-arg ldiv!
     @eval ldiv!(out::$Tri, D::Diagonal, A::$Tri) = $Tri(ldiv!(out.data, D, A.data))
-    @eval ldiv!(out::$Tri, D::Diagonal, A::$UTri) = $Tri(_set_diag!(ldiv!(out.data, D, A.data), D.diag, inv))
-    @eval mul!(out::$Tri, D::Diagonal, A::$Tri) = $Tri(mul!(out.data, D, A.data))
-    @eval mul!(out::$Tri, D::Diagonal, A::$UTri) = $Tri(_set_diag!(mul!(out.data, D, A.data), D.diag))
-    @eval mul!(out::$Tri, A::$Tri, D::Diagonal) = $Tri(mul!(out.data, A.data, D))
-    @eval mul!(out::$Tri, A::$UTri, D::Diagonal) = $Tri(_set_diag!(mul!(out.data, A.data, D), D.diag))
+    @eval ldiv!(out::$Tri, D::Diagonal, A::$UTri) = $Tri(_setdiag!(ldiv!(out.data, D, A.data), inv, D.diag))
+    # 3-arg mul!: invoke 5-arg mul! rather than lmul!
+    @eval mul!(out::$Tri, A::Union{$Tri,$UTri}, D::Diagonal) = mul!(out, A, D, true, false)
+    # 5-arg mul!
+    @eval @inline mul!(out::$Tri, D::Diagonal, A::$Tri, α::Number, β::Number) =
+        $Tri(mul!(out.data, D, A.data, α, β))
```

Comment:
Since 
```julia
julia> let a = randn(10,10)
           UpperTriangular(a) === UpperTriangular(a)
       end
true
```
I think it's OK? And `rmul!` was implemented as:
```julia
rmul!(A::Union{LowerTriangular,UpperTriangular}, D::Diagonal) = typeof(A)(rmul!(A.data, D))
```
I'll add these tests anyway.